actioncable/app/javascript/action_cable/action_cable.js

```diff
@@ -0,0 +1,43 @@
+import Connection from "./connection"
+import ConnectionMonitor from "./connection_monitor"
+import Consumer from "./consumer"
+import INTERNAL from "./internal"
+import { log, logger } from "./logger"
+import Subscription from "./subscription"
+import Subscriptions from "./subscriptions"
+
+export function getConfig(name) {
+  const element = document.head.querySelector(`meta[name='action-cable-${name}']`)
+  return (element ? element.getAttribute("content") : undefined)
+}
+
+export function createWebSocketURL(url) {
+  if (url && !/^wss?:/i.test(url)) {
+    const a = document.createElement("a")
+    a.href = url
+    // Fix populating Location properties in IE. Otherwise, protocol will be blank.
+    a.href = a.href
+    a.protocol = a.protocol.replace("http", "ws")
+    return a.href
+  } else {
+    return url
+  }
+}
+
+export function createConsumer(url) {
+  if (url == null) {
+    const urlConfig = getConfig("url")
+    url = (urlConfig ? urlConfig : INTERNAL.default_mount_path)
+  }
+  return new Consumer(createWebSocketURL(url))
+}
+
+export function startDebugging() {
+  logger.enabled = true
+}
+
+export function stopDebugging() {
+  logger.enabled = false
+}
+
+export { Connection, ConnectionMonitor, Consumer, INTERNAL, log, Subscription, Subscriptions }
```

Comment:
Follow up to https://github.com/rails/rails/pull/34258#discussion_r229386444:

Ha, that's a pretty bad test! It just confirms that yes in fact JavaScript does allow assigning expando properties wherever you'd like. What it doesn't test, and what I'm pretty sure won't work, is actually using the assigned adapter here: https://github.com/rails/rails/blob/629b2d6333451cf751bceb5e0e834a6e1e1c15c3/actioncable/app/javascript/action_cable/connection.js#L38

Reply:
Unless I'm missing something? I don't see the adapter interface being exported here.

Reply:
Ah yeah, that's a bad test (but it does fail if I remove the setter defined here: https://github.com/rails/rails/blob/629b2d6333451cf751bceb5e0e834a6e1e1c15c3/actioncable/app/javascript/action_cable/index.js#L12).

Setting the adapter is implicitly tested via the `consumerTest` test cases, because `consumerTest` sets `ActionCable.WebSocket` here: https://github.com/rails/rails/blob/405aef3dede8938c8eb9789f6d6722864faddb09/actioncable/test/javascript/src/test_helpers/consumer_test_helper.coffee#L15

If I remove the setter referenced above, then the tests declared with `consumerTest` fail.

Here's a more direct test of the line you referenced (not part of the current test suite): https://github.com/rails/rails/blob/4489cd5b03a58d062ca51f21e41092d28652c7cd/actioncable/test/javascript/src/unit/connection_test.coffee#L1-L13

It passes with no changes and fails if I remove the setter referenced above.

Reply:
Ah, my bad! I totally missed that there were separate [`index.js`](https://github.com/rails/rails/blob/dc5c137985943b7a39e99e02c78a7918e8a90833/actioncable/app/javascript/action_cable/index.js) and [`action_cable.js`](https://github.com/rails/rails/blob/dc5c137985943b7a39e99e02c78a7918e8a90833/actioncable/app/javascript/action_cable/action_cable.js) files. I would prefer they be combined so that `index.js` is the main (and only) export for the `ActionCable` interface.

Reply:
I've been following this PR out of excitement for this change and I wanted to add my thoughts here. I was not too fond of the original approach that had `index.js` and a separate `index_with_api_bridge.js` which was rather messy. But I very much like the simpler `index.js` and `action_cable.js` split (IMO the latter could be renamed to just `cable.js` though).

I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.

Even with `index.js` and `cable.js` as separate files, the `index.js` entry point is the main export. It has the entire `ActionCable` interface, including compatibility with the `ActionCable.WebSocket = ...` and `ActionCable.logger = ...` interface from ActionCable v5.x, and it's what you get when you simply `import ActionCable from "actioncable"` (or use `const ActionCable = require("ActionCable")`, or depend on the global variable from the UMD module). That's excellent.

But by also having the separate `cable.js` file with individually exported functions, we're giving users the ability to import `createConsumer` directly and avoid shipping everything else that's not used in their application. I think that's worth having the additional source file.

For those of us working on an application with a [performance budget](https://infrequently.org/2017/10/can-you-afford-it-real-world-web-performance-budgets/), shaving even a few KB off our dependencies gives us considerably more room to grow our application's code. This is why people like @developit and @lukeed have spent lots of time refactoring their libraries to shave off a couple hundred bytes here and there. I'm not suggesting ActionCable should worry to that extent. But I can't see why not to offer the additional tree-shaking friendly savings that come for free if we keep `index.js` and `cable.js` separate.

Reply:
I agree that we shouldn't sacrifice the tree-shaking benefits.

However, if we don't want to have the two separate source files, there is an alternative that I would be fine with: We could also export `adapters` from `action_cable.js` and move `action_cable.js` to `index.js`. Then all functionality would be available from the individual exports of the new `index.js` (former `action_cable.js`), but we would be slightly changing the public API. I don't think that's a huge cause for concern because:
- We can share the contents of what is currently `index.js` in the upgrade guide as a way for users to patch the old API into their applications when they upgrade, until they migrate to the new API.
- We could go further and ship a codemod that they can run to automatically translate their application to the new API. (If you're not familiar, this is something [react has been doing](https://github.com/reactjs/react-codemod) to make API upgrades easy for users.) I'd be happy to look into writing the codemod.

Reply:
> I support keeping `index.js` and `cable.js` separate because if we combined them, we'd lose the ability for applications to tree-shake unused functions from ActionCable. I don't see why we should sacrifice that in order to have one fewer source file.

Can you share an example of that? How would you cherry-pick your imports for optimal tree-shaking with those separate files and what would the difference be if they were combined?

Reply:
For this PR, I'd prefer to have everything in `index.js` so there's a single export point. That way it's a clear, straightforward record of the conversion from CoffeeScript to plain JS.

Then, if you'd like, follow up with a second PR that splits the files more optimally and documents the before/after tree-shaking improvements.

Reply:
Ok, I’ll move the changes that are not the direct CoffeeScript conversion back to https://github.com/rails/rails/pull/34258.

Reply:
Done. This PR is now only including the CoffeeScript to ES2015 translation and there's only `index.js`. Let's proceed with getting this one merged now and continue further module structure discussion at https://github.com/rails/rails/pull/34258#discussion_r229372522.

Reply:
Thanks. FWIW, I'm +1 on your change to export individual functions like `export function createConsumer()` instead of `export const ActionCable`, and would be happy to have that included in this PR. I'd just like the `logger` and `WebSocket` interface to be similarly exported, and from the same file (`index.js`).

Reply:
Since there's no separate `action_cable.js` in this PR anymore, and we can evaluate the benefits of having that file in #34258, I'm marking this thread resolved.